From 8c526ede5039ecdeb2f39fff3be902d73609dc57 Mon Sep 17 00:00:00 2001
From: Andrew Wang <wang@apache.org>
Date: Wed, 27 Apr 2016 17:13:50 -0700
Subject: [PATCH 1470/2748] HADOOP-13030. Handle special characters in
 passwords in KMS startup script. Contributed by
 Xiao Chen.

(cherry picked from commit 7271e91b791ba277fd2a9b8f870eb9bc4f395566)
(cherry picked from commit 0d4fbf01beb6d4ac853f1fd7d74d735c1d0c0561)

Conflicts:
	hadoop-common-project/hadoop-kms/src/main/sbin/kms.sh

Change-Id: Iba1e94686bd4496dd1666487e116b0c3b1381237
---
 .../hadoop-kms/src/main/sbin/kms.sh                |   14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/hadoop-common-project/hadoop-kms/src/main/sbin/kms.sh b/hadoop-common-project/hadoop-kms/src/main/sbin/kms.sh
index f0cb158..f5b9d0a 100644
--- a/hadoop-common-project/hadoop-kms/src/main/sbin/kms.sh
+++ b/hadoop-common-project/hadoop-kms/src/main/sbin/kms.sh
@@ -26,6 +26,14 @@ while [ -h "${PRG}" ]; do
   fi
 done
 
+function hadoop_escape() {
+      # Escape special chars for the later sed which saves the text as xml attribute
+      local ret
+      ret=$(sed 's/[\/&]/\\&/g' <<< "$1" | sed 's/&/\&amp;/g' | sed 's/"/\\\&quot;/g' \
+          | sed "s/'/\\\\\&apos;/g" | sed 's/</\\\&lt;/g' | sed 's/>/\\\&gt;/g')
+      echo "$ret"
+}
+
 BASEDIR=`dirname ${PRG}`
 BASEDIR=`cd ${BASEDIR}/..;pwd`
 
@@ -76,9 +84,11 @@ fi
 if [ ! "${KMS_SSL_KEYSTORE_PASS}" = "" ] || [ ! "${KMS_SSL_TRUSTSTORE_PASS}" = "" ]; then
   # Set a KEYSTORE_PASS if not already set
   KMS_SSL_KEYSTORE_PASS=${KMS_SSL_KEYSTORE_PASS:-password}
+  KMS_SSL_KEYSTORE_PASS_ESCAPED=$(hadoop_escape "$KMS_SSL_KEYSTORE_PASS")
+  KMS_SSL_TRUSTSTORE_PASS_ESCAPED=$(hadoop_escape "$KMS_SSL_TRUSTSTORE_PASS")
   cat ${CATALINA_BASE}/conf/ssl-server.xml.conf \
-    | sed 's/_kms_ssl_keystore_pass_/'${KMS_SSL_KEYSTORE_PASS}'/g' \
-    | sed 's/_kms_ssl_truststore_pass_/'${KMS_SSL_TRUSTSTORE_PASS}'/g' > ${CATALINA_BASE}/conf/ssl-server.xml
+    | sed 's/"_kms_ssl_keystore_pass_"/'"\"${KMS_SSL_KEYSTORE_PASS_ESCAPED}\""'/g' \
+    | sed 's/"_kms_ssl_truststore_pass_"/'"\"${KMS_SSL_TRUSTSTORE_PASS_ESCAPED}\""'/g' > ${CATALINA_BASE}/conf/ssl-server.xml
   cp ${CATALINA_BASE}/conf/ssl-server.xml ${CATALINA_BASE}/conf/server.xml
 else
   cp ${CATALINA_BASE}/conf/server.xml.conf ${CATALINA_BASE}/conf/server.xml
-- 
1.7.9.5

